### ISSUE TITLE
RuntimeError: derivative for dequantize is not implemented

### ISSUE THREAD

**Issue by Eugene29** [huggingface/optimum-quanto#212]
Hi, first of all, thanks for creating this package. This is awesome, so much better than the one pytorch officially offers. 

I want to try QAT, but I'm getting the following error: 
```
[rank3]: Traceback (most recent call last):
[rank3]:   File "/home/euku/aoboResearch/long_range_GNN/pointnet/cleaned_pointNET/Quantization.py", line 171, in <module>
[rank3]:     loss.backward()
[rank3]:   File "/home/euku/stable/lib/python3.9/site-packages/torch/_tensor.py", line 525, in backward
[rank3]:     torch.autograd.backward(
[rank3]:   File "/home/euku/stable/lib/python3.9/site-packages/torch/autograd/__init__.py", line 267, in backward
[rank3]:     _engine_run_backward(
[rank3]:   File "/home/euku/stable/lib/python3.9/site-packages/torch/autograd/graph.py", line 744, in _engine_run_backward
[rank3]:     return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
[rank3]: RuntimeError: derivative for dequantize is not implemented
```

My code looks like this: 
```
from optimum.quanto import qint8
from optimum.quanto.calibrate import Calibration
from optimum.quanto.quantize import quantize, freeze
quantize(model, weights=qint8, activations=qint8)
with torch.no_grad():
    with Calibration(momentum=0.9):
        for batch in train_loader:
            X, y = batch
            model(X)
if args.QAT:
    model.train()
    current_lr = optimizer.param_groups[0]['lr']
    print(current_lr)
    for epoch, batch in enumerate(train_loader):
        optimizer.zero_grad()
        X, y = batch
        out = model(X).dequantize()

        ## loss
        xyz_loss = F.mse_loss(out[:, :-1], y[:, :-1])
        energy_loss = args.scale_energy_loss * F.l1_loss(out[:, -1], y[:, -1])
        loss = xyz_loss + energy_loss

        ## gradient
        loss.backward()
        optimizer.step()

freeze(model)
```

**Comment by dacorvo** [comment]
I think your model output is already dequantized, and thus a regular torch.Tensor.
As a consequence:
- when calling dequantize(), you actually call the pytorch no-op for non-quantized tensors,
- you add the no-op to the graph but it does not implement backward() (which explains the message about no derivative being defined).
If you look at the MNIST example, there is an extra check before calling dequantize:
```
if isinstance(output, QTensor):
    output = output.dequantize()
```
Note that I think this is now useless in the MNIST example, as when the loss for that `QTensor` is evaluated it should be dequantized implicitly.
Aniyway, I think it will work if you remove the call to dequantize.

**Comment by Eugene29** [comment]
Oh Thanks! Your explanation makes total sense and getting rid of dequantize fixed the problem.

**Comment by LukeLIN-web** [comment]
> Hi, first of all, thanks for creating this package. This is awesome, so much better than the one pytorch officially offers.
> 
> I want to try QAT, but I'm getting the following error:
> 
> ```
> [rank3]: Traceback (most recent call last):
> [rank3]:   File "/home/euku/aoboResearch/long_range_GNN/pointnet/cleaned_pointNET/Quantization.py", line 171, in <module>
> [rank3]:     loss.backward()
> [rank3]:   File "/home/euku/stable/lib/python3.9/site-packages/torch/_tensor.py", line 525, in backward
> [rank3]:     torch.autograd.backward(
> [rank3]:   File "/home/euku/stable/lib/python3.9/site-packages/torch/autograd/__init__.py", line 267, in backward
> [rank3]:     _engine_run_backward(
> [rank3]:   File "/home/euku/stable/lib/python3.9/site-packages/torch/autograd/graph.py", line 744, in _engine_run_backward
> [rank3]:     return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
> [rank3]: RuntimeError: derivative for dequantize is not implemented
> ```
> 
> My code looks like this:
> 
> ```
> from optimum.quanto import qint8
> from optimum.quanto.calibrate import Calibration
> from optimum.quanto.quantize import quantize, freeze
> quantize(model, weights=qint8, activations=qint8)
> with torch.no_grad():
>     with Calibration(momentum=0.9):
>         for batch in train_loader:
>             X, y = batch
>             model(X)
> if args.QAT:
>     model.train()
>     current_lr = optimizer.param_groups[0]['lr']
>     print(current_lr)
>     for epoch, batch in enumerate(train_loader):
>         optimizer.zero_grad()
>         X, y = batch
>         out = model(X).dequantize()
> 
>         ## loss
>         xyz_loss = F.mse_loss(out[:, :-1], y[:, :-1])
>         energy_loss = args.scale_energy_loss * F.l1_loss(out[:, -1], y[:, -1])
>         loss = xyz_loss + energy_loss
> 
>         ## gradient
>         loss.backward()
>         optimizer.step()
> 
> freeze(model)
> ```

I am curious why it can achieve QAT, because in my understanding, QAT need update zero point and scale in every iteration, but how you achieve this? Can backward update zero point and scale?

**Comment by dacorvo** [comment]
@LukeLIN-web thank you for the nice feedback.
TLDR: The reason why you get the error is because the output of your model is already dequantized.
Longer explaination:
- when activations are quantized, the inputs/outputs of every Linear or Conv layer are expressed as QTensor, a subclass of Tensor (that takes less m=device memory obviously),
- since not all operations are supported on this subclass, the QTensor can be dequantized into a Tensor, either explicitely or automatically, and **this operation is differentiable**,
- in the training loop, the evaluation of the loss on a QTensor is not possible, so it has to be converted to a Tensor first. This is why in the MNIST example I added this call to dequantize,
- however, and I confess this is utterly confuzing, the Tensor class also has a dequantize method (comes from the legacy support for quantizaiton in pytorch), and this one **is not differentiable**,
- this is why you need to call dequantize only if the output of your model is a QTensor (see https://github.com/huggingface/optimum-quanto/blob/66dca87bc7e61ffbf1a06da9fd324469592e17e9/examples/vision/image-classification/mnist/quantize_mnist_model.py#L72C9-L72C40/).

**Comment by dacorvo** [comment]
Alternatively, I think that you can safely omit the call to dequantize since quanto will automatically dequantize the QTensor for you when it reaches an unsupported operation.

**Comment by dacorvo** [comment]
But I realize I had already provided the same answer back in june ...  😅

**Comment by LukeLIN-web** [comment]
> But I realize I had already provided the same answer back in june ... 😅

Thank you for your detailed explanation, it makes us understand it more clearly.
